### PR TITLE
Promises

### PR DESCRIPTION
**Pull request by julianhille**
The idea behind the config variable injected into every test is that i like to
inject my database object programatically otherwise i've some kind of cyclomatic imports / requires or code duplication. As im working with entity services / models which "hides" the database  i would need 
to export the database ONLY for migrations.
Its also easier to test the migrations when im able to inject my "mocked" database into the migration
functions without the need to mock a lot of other objects too.

This is something like a circle of imports:
* initiate database
* start migration
* migrations import `initiate database`

There is a lot of room for improvements:
- Tests could be in a lot of cases simplified.
- I miss some tests for the new config variable passed to every test
- allow "config" argument for the CLI so that one is able to use the config also for manually triggered migrations

### CODE REVIEW

File: test/util/run.js [hunk 1]
```diff
@@ -2,18 +2,20 @@
...
-  })
-  p.stderr.on('data', function (d) {
-    out += d.toString('utf8')
-  })
-  p.on('error', done)
-  p.on('close', function (code) {
-    done(null, out, code)
+var run = module.exports = async function run (cmd, dir, args, done) {
+  return new Promise((resolve, reject) => {
+    var p = spawn(cmd, ['-c', dir, ...args])
```
Comment (LinusU) [inline test/util/run.js]: Have you looked at [execa](https://github.com/sindresorhus/execa)?

File: test/issue-33.js [hunk 2]
```diff
@@ -12,45 +12,32 @@ const STATE = path.join(BASE, '.migrate')
...
-    })
+    await set.down()
+    assert.deepEqual(db.issue33, A2)
   })
 
-  afterEach(function (done) {
+  afterEach(function () {
     db.nuke()
-    fs.unlink(STATE, done)
+    fs.unlinkSync(STATE)
```
Comment (LinusU) [inline test/issue-33.js]:
```
const fs = require('fs')
const util = require('util')

// ...

const unlink = util.promisify(fs.unlink)

// ...

await unlink(STATE)
```

**Comment by LinusU** [comment]
I haven't had time to look at this properly, but some quick thoughts:

1) I don't think we can use `async/await` since it would require Node.js 8, I think we should strive to, at the very least, support all LTS versions of Node.

2) Seems like a better approach might be to do this incrementally, and start with all the internal stuff before breaking the external API, that way we could hold of with the breaking changes but still benefit from promisified internals.

3) I love promises and async/await 😍

**Comment by julianhille** [comment]
> 1.  don't think we can use async/await since it would require Node.js 8, I think we should strive to, at the very least, support all LTS versions of Node.

It landed in node 7.6 with the V8 engine. 

> 2. Seems like a better approach might be to do this incrementally, and start with all the internal stuff before breaking the external API, that way we could hold of with the breaking changes but still benefit from promisified internals.

That's why i upgraded the major version.

> 3. I love promises and async/await

Me too.

I can not wait for Node 6 get out of date around arpil 2019 so i guess i'll do a hard fork then.

File: test/issue-33.js (hunk 2 above)
Comment (julianhille) [reply test/issue-33.js]: i could have used bluebird too, but this is in a test and was not of a major point for me to solve that. Support my point about loads of improvements are out there.

File: test/util/run.js (hunk 1 above)
Comment (julianhille) [reply test/util/run.js]: Did not know this. Would leave this for another PR.

File: test/issue-33.js (hunk 2 above)
Comment (LinusU) [reply test/issue-33.js]: `util.promisify` is built into Node 🙌 no need to install any libraries

Comment (julianhille) [reply test/issue-33.js]: True. Actually added bluebird to use it for another purpose which was solved differently, i like the ease of `promisifyAll` but it adds a bunch of dependencies.

**Comment by wesleytodd** [comment]
I also haven't had a chance to dig in to this PR, but IMO we should support all versions in active maintenance mode.  These new features do not improve the end user experience for this library, so I am hesitant to make this change now.  

I would like to dig more into the actual changes before we decide how to move forward, but lets let this sit atleast until this weekend when we can all have more time to look at it.  Sound alright?

File: lib/log.js [hunk 3]
```diff
@@ -7,4 +7,5 @@ module.exports = function log (key, msg) {
 
 module.exports.error = function log (key, msg) {
   console.error('  ' + chalk.red(key) + ' : ' + chalk.white(msg))
+  console.trace()
```
Comment (julianhille) [inline lib/log.js]: debug output, need to remove.

File: bin/migrate-list [hunk 4]
```diff
@@ -13,66 +13,78 @@ var registerCompiler = require('../lib/register-compiler')
...
+.option('--env [name]', 'Use dotenv to load an environment file')
+.parse(process.argv)
 
-// Check clean flag, exit if NODE_ENV === 'production' and force not specified
-if (program.clean && process.env.NODE_ENV === 'production' && !program.force) {
-  log.error('error', 'Cowardly refusing to clean while node environment set to production, use --force to continue.')
-  process.exit(1)
-}
+async function list (params) {
+  console.log('calling me', params)
```
Comment (julianhille) [inline bin/migrate-list]: debug output, need to be removed.

File: bin/migrate-list [hunk 5]
```diff
@@ -13,66 +13,78 @@ var registerCompiler = require('../lib/register-compiler')
...
-    return log('list', 'No Migrations')
+    set.migrations.forEach(function (migration) {
+      let timeString = ' [not run] '
+      if (migration.timestamp) {
+        timeString = ` [${dateFormat(migration.timestamp, program.dateFormat)}]`
+      }
+      log(`${migration.title} ${timeString}`, migration.description || '<No Description>')
+    })
+  } catch (error) {
+    console.log(error)
```
Comment (julianhille) [inline bin/migrate-list]: debug output, need to be removed

**Comment by jmjpro** [comment]
As an aside, node 4 support ends the end of April 2018 (next month): https://github.com/nodejs/Release. This means that node-migrate could very soon start using a bunch of features introduced between node 4 and node 6 such as default function parameters, rest parameters, spread operator, destructuring, Array.prototype.filter/map/includes, and many more...

**Pull request closed without merging**